### PR TITLE
[fast] if cluster is INIT, force refresh before deciding to provision

### PR DESCRIPTION
**Pull request by cg505**
If a cluster is mid-initialization, its status will be INIT and autostop/down will not be set yet. In this case, the cluster refresh won't actually grab the cluster status lock and hard refresh the status. So, check_cluster_available will immeidately decide that the cluster is INIT and throw.

This could cause a bug where many parallel launches of `sky launch --fast` that are staggered can all decide that the cluster is INIT, and all decide that they need to launch the cluster. Since cluster initialization is locked with the cluster status lock, each invocation will sychronously re-launch the cluster.

Now, if we see that the cluster is INIT, we force a refresh. This will acquire the cluster status lock, which will block until any ongoing provisioning completes and the cluster is UP. If the cluster is otherwise INIT (e.g. ray cluster has been stopped abnormally) then provisioning should proceed as normal.

This does not fix the race where the cluster does not exist or is STOPPED, and many simultaneously started `sky launch --fast` invocations try to create or restart the cluster. However, once the first batch complete their launches, all future invocations should correctly see the cluster as UP, not INIT - even if they are started while the first batch is still provisioning the cluster. Fixing the STOPPED or non-existent case is a bit more difficult and will probably require moving this detection logic inside the provisioning code, so that it holds the cluster status lock continuously from the status check until the cluster is UP.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @cg505!

File: sky/execution.py [hunk 1]
```diff
@@ -463,28 +463,34 @@ def launch(
...
-            try:
-                # This will throw if the cluster is not available
-                backend_utils.check_cluster_available(
+        cluster_status, maybe_handle = (
+            backend_utils.refresh_cluster_status_handle(cluster_name))
+        if cluster_status == status_lib.ClusterStatus.INIT:
+            # If the cluster is provisioning we should actually hard refresh the
+            # state with the appropriate lock.
+            cluster_status, maybe_handle = (
+                backend_utils.refresh_cluster_status_handle(
```
[the remark this answers is not in the document]
Comment (cg505) [reply sky/execution.py]: Expanded the comment to explain this.

[the remark this answers is not in the document]
Comment (cg505) [reply sky/execution.py:481]: Added this to the comment.

**Review by romilbhardwaj — APPROVED** [review]
Awesome, thanks @cg505!

**Review by romilbhardwaj — APPROVED** [review]
Awesome, thanks @cg505!

**Pull request merged**